api4/websocket.go

```diff
@@ -16,23 +20,47 @@ func (api *API) InitWebSocket() {
 }
 
 func connectWebSocket(c *Context, w http.ResponseWriter, r *http.Request) {
-	upgrader := websocket.Upgrader{
-		ReadBufferSize:  model.SOCKET_MAX_MESSAGE_SIZE_KB,
-		WriteBufferSize: model.SOCKET_MAX_MESSAGE_SIZE_KB,
-		CheckOrigin:     c.App.OriginChecker(),
+	fn := c.App.OriginChecker()
+	if fn != nil && !fn(r) {
+		c.Err = model.NewAppError("origin_check", "api.web_socket.connect.check_origin.app_error", nil, "", http.StatusBadRequest)
+		return
+	}
+
+	upgrader := ws.HTTPUpgrader{
+		Timeout: 5 * time.Second,
 	}
 
-	ws, err := upgrader.Upgrade(w, r, nil)
+	conn, _, _, err := upgrader.Upgrade(r, w)
 	if err != nil {
 		c.Err = model.NewAppError("connect", "api.web_socket.connect.upgrade.app_error", nil, "", http.StatusInternalServerError)
 		return
 	}
 
-	wc := c.App.NewWebConn(ws, *c.App.Session(), c.App.T, "")
-
+	wc := c.App.NewWebConn(conn, *c.App.Session(), c.App.T, "")
 	if len(c.App.Session().UserId) > 0 {
 		c.App.HubRegister(wc)
 	}
 
-	wc.Pump()
+	go wc.Pump()
+
+	desc := netpoll.Must(netpoll.HandleRead(conn))
+	c.App.Srv().Poller().Start(desc, func(wsEv netpoll.Event) {
+		if wsEv&(netpoll.EventReadHup|netpoll.EventHup) != 0 {
+			c.App.Srv().Poller().Stop(desc)
+			wc.Close()
+			return
+		}
+
+		// read from conn
+		// TODO(agniva): Implement semaphore to force backpressure
+		// on goroutine creation.
+		go func() {
+			err := wc.ReadMsg()
+			if err != nil {
+				mlog.Debug("Error while reading message from websocket", mlog.Err(err))
+				c.App.Srv().Poller().Stop(desc)
+				wc.Close()
+			}
+		}()
+	})
```

Comment:
Should this be part of `wc.Pump()`? It feels strange to have such a low level logic in the api package.

Reply:
Yeah, I'll see what I can do.

Reply:
Done